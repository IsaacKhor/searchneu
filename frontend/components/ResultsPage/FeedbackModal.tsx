import React, { useState, useEffect, useRef } from 'react';
import LogoInput from '../images/LogoInput';
import CheckboxGroup from './CheckboxGroup';
import macros from '../macros';
import useFeedbackSchedule from './useFeedbackSchedule';

interface InitialStepProps {
  yes: boolean;
  setYes: (b:boolean) => void;
}


function FeedbackModalInitial({ yes, setYes } : InitialStepProps) {
  return (
    <>
      <div className='FeedbackModal__popoutSubHeader'>
        <p>Did you find what you were looking for?</p>
      </div>
      <div className='FeedbackModal__initial'>
          <label className='FeedbackModal_initialText' onClick={() => setYes(true)}>
            <input type='radio' name='FeedbackRadio'/>
            <span className='FeedbackModal__initialRadio' />
            Yes
          </label>
          <label className='FeedbackModal__initialText' onClick={() => setYes(false)}>
            <input type='radio' name='FeedbackRadio'/>
            <span className='FeedbackModal__initialRadio' />
            No
          </label>
      </div>
    </>


  );
}

interface CheckboxProps {
  feedbackOptions: string[];
  selectedFeedback: string[];
  setSelectedFeedback: (a:string[]) => void;
}

function FeedbackModalCheckboxes({ feedbackOptions, selectedFeedback, setSelectedFeedback } : CheckboxProps) {
  return (
    <>
      <div className='FeedbackModal__popoutSubHeader'>
        <p>What info are you looking for?</p>
      </div>
      <div className='FeedbackModal__checkBoxes'>
        <CheckboxGroup name='FeedbackModalCheckboxes' value={ selectedFeedback } onChange={ setSelectedFeedback }>
          {(Checkbox) => (
            <>
              {feedbackOptions.map((feedbackOption) => (
                <div key={ feedbackOption } className='FeedbackModal__checkboxElement'>
                  <label className='FeedbackModal__checkboxText'>
                    <Checkbox value={ feedbackOption } />
                    <span className='FeedbackModal__checkboxBox' />
                    {feedbackOption}
                  </label>
                </div>
              ))}
            </>
          )}
        </CheckboxGroup>
      </div>
    </>
  );
}

export default function FeedbackModal() {
  const [open, setOpen] = useState(false);
  const [yes, setYes] = useState(false);
  const [selectedFeedback, setSelectedFeedback] = useState([]);
  const [submitted, setSubmitted] = useState(false);
  const [step, setStep] = useState('initial');
  const modalRef = useRef(null);
  const keyString = 'MODAL'
  const [show, setFinished] = useFeedbackSchedule(keyString, 86400000);

  const feedbackOptions = ['Class time', 'Professor', 'Prereqs', 'Something else'];

  const handleClickOutside = (e) => {
    if (modalRef.current.contains(e.target)) {
      return;
    }
    setOpen(false);
  }
  console.log('yes', yes);

  const handleSubmit = () => {
    switch (step) {
      case 'initial':
        if (yes) {
          macros.logAmplitudeEvent('Feedback modal initial submit', {lookingForFound: yes});

        } else {
          macros.logAmplitudeEvent('Feedback modal initial submit', {lookingForFound: yes});
          setStep('checkbox');

        }
      case 'checkbox':
          macros.logAmplitudeEvent('Feedback modal checkbox submit', { lookingFor: selectedFeedback }); };

        
  }


  const renderFeedback = () => {
    switch (step) {
      case 'initial':
        return <FeedbackModalInitial yes={ yes } setYes={ setYes } />;
      case 'checkbox':
        return <FeedbackModalCheckboxes feedbackOptions={ feedbackOptions } selectedFeedback={ feedbackOptions } setSelectedFeedback={ setSelectedFeedback } />;
    }
  }

  useEffect(() => {
    if (open) {
      document.addEventListener('mousedown', handleClickOutside);
    } else {
      document.removeEventListener('mousedown', handleClickOutside);
    }
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
    };
  }, [open]);


  return (
    show
    && (
    <div ref={ modalRef } className='FeedbackModal'>
      {open && (
      <div className='FeedbackModal__popout'>
        <div className='FeedbackModal__popoutHeader'>
          <p>SearchNEU Feedback</p>
        </div>
        { renderFeedback()}
        <div className={ !submitted ? 'FeedbackModal__submit' : 'FeedbackModal__submit--submitted' } role='button' tabIndex={ 0 } onClick={ handleSubmit }>
          <p>{!submitted ? 'SEND FEEDBACK' : 'THANK YOU!' }</p>
        </div>
      </div>
      ) }
      <div className='FeedbackModal__pill' role='button' tabIndex={ 0 } onClick={ () => { setOpen(!open) } }>
        <LogoInput height='14' width='18' fill='#d41b2c' />
        <p>SearchNEU Feedback</p>
      </div>
    </div>
    )


  );
}
