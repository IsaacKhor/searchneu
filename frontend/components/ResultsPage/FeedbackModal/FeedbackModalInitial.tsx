import React from 'react';

interface InitialStepProps {
  setYes: (b:boolean) => void;
}

export default function FeedbackModalInitial({ setYes } : InitialStepProps) {
  return (
    <>
      <div className='FeedbackModal__popoutSubHeader'>
        <p>Did you find what you were looking for?</p>
      </div>
      <div className='FeedbackModal__initial'>
        <label className='FeedbackModal__initialText' onClick={ () => setYes(true) }>
          <input type='radio' name='FeedbackRadio' />
          <span className='FeedbackModal__initialRadio' />
          Yes
        </label>
        <label className='FeedbackModal__initialText' onClick={ () => setYes(false) }>
          <input type='radio' name='FeedbackRadio' />
          <span className='FeedbackModal__initialRadio' />
          No
        </label>
      </div>
    </>
  );
}
