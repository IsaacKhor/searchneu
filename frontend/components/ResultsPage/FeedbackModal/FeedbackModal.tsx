import React, { useState, useRef } from 'react';
import LogoInput from '../../images/LogoInput';
import FeedbackModalInitial from './FeedbackModalInitial';
import FeedbackModalCheckboxes from './FeedbackModalCheckboxes';
import FeedbackModalFree from './FeedbackModalFree';
import macros from '../../macros';
import useFeedbackSchedule from '../useFeedbackSchedule';
import useClickOutside from '../useClickOutside';


export default function FeedbackModal() {
  const [open, setOpen] = useState(false);
  const [yes, setYes] = useState(false);
  const [selectedFeedback, setSelectedFeedback] = useState([]);
  const [feedbackQuery, setFeedbackQuery] = useState('');
  const [feedbackType, setFeedbackType] = useState('filter');
  const [submitted, setSubmitted] = useState(false);
  const [step, setStep] = useState('initial');

  const modalRef = useRef(null);
  useClickOutside(modalRef, open, setOpen);

  const keyString = 'MODAL'
  const [show, setFinished] = useFeedbackSchedule(keyString, 86400000);

  const feedbackOptions = ['Class time', 'Professor', 'Prereqs', 'Something else'];

  function handleSubmit() {
    switch (step) {
      case 'initial':
        if (yes) {
          macros.logAmplitudeEvent('Feedback modal initial submit', { lookingForFound: yes });
          setSubmitted(true);
          setFinished();
        } else {
          macros.logAmplitudeEvent('Feedback modal initial submit', { lookingForFound: yes });
          setStep('checkbox');
        }
        break;
      case 'checkbox':
        macros.logAmplitudeEvent('Feedback modal checkbox submit', { lookingFor: selectedFeedback });
        setStep('free');
        break;
      case 'free':
        macros.logAmplitudeEvent('Feedback modal free submit', { feedbackType: feedbackType, feedbackQuery: feedbackQuery });
        setSubmitted(true);
        setFinished();
        break;
      default:
        break;
    }
  }

  function buttonText() {
    if (yes && submitted) {
      return 'HOLLA HOLLA';
    } if (submitted) {
      return 'THANK YOU!';
    }
    return 'SEND FEEDBACK'
  }

  function renderFeedback() {
    switch (step) {
      case 'initial':
        return <FeedbackModalInitial setYes={ setYes } />;
      case 'checkbox':
        return <FeedbackModalCheckboxes feedbackOptions={ feedbackOptions } selectedFeedback={ selectedFeedback } setSelectedFeedback={ setSelectedFeedback } />;
      case 'free':
        return <FeedbackModalFree feedbackQuery={ feedbackQuery } setFeedbackQuery={ setFeedbackQuery } setFeedbackType={ setFeedbackType } />;
      default:
        return null;
    }
  }

  return (
    show
    && (
    <div ref={ modalRef } className='FeedbackModal'>
      {open && (
      <div className='FeedbackModal__popout'>
        <div className='FeedbackModal__popoutHeader'>
          <p>SearchNEU Feedback</p>
        </div>
        { renderFeedback()}
        <div className={ !submitted ? 'FeedbackModal__submit' : 'FeedbackModal__submit--submitted' } role='button' tabIndex={ 0 } onClick={ handleSubmit }>
          <p>{buttonText()}</p>
        </div>
      </div>
      ) }
      <div className='FeedbackModal__pill' role='button' tabIndex={ 0 } onClick={ () => { setOpen(!open) } }>
        <LogoInput height='14' width='18' fill='#d41b2c' />
        <p>SearchNEU Feedback</p>
      </div>
    </div>
    )
  );
}
