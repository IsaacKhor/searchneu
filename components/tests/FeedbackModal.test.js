/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import React from 'react';

import Enzyme, { shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import FeedbackModal from '../FeedbackModal';

Enzyme.configure({ adapter: new Adapter() });

it('should render the form', () => {
  const result = shallow(
    <FeedbackModal
      closeForm={Function}
      feedbackModalOpen
      toggleForm={Function}
    />
  ).debug();
  expect(result).toMatchSnapshot();
});

it('should render form is closed', () => {
  const result = shallow(
    <FeedbackModal
      closeForm={Function}
      feedbackModalOpen={false}
      toggleForm={Function}
    />
  ).debug();
  expect(result).toMatchSnapshot();
});
