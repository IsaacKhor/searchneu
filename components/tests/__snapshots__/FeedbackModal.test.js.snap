// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render form is closed 1`] = `
"<div className=\\"feedback-container\\">
  <Transition in={false} timeout={500} mountOnEnter={false} unmountOnExit={false} appear={false} enter={true} exit={true} onEnter={[Function: noop]} onEntering={[Function: noop]} onEntered={[Function: noop]} onExit={[Function: noop]} onExiting={[Function: noop]} onExited={[Function: noop]}>
    [function]
  </Transition>
  <Modal open={false} onClose={[Function: Function]} size=\\"small\\" className=\\"feedback-modal-container\\" centered={true} dimmer={true} closeOnDimmerClick={true} closeOnDocumentClick={false} eventPool=\\"Modal\\">
    <Header icon=\\"mail\\" content=\\"Get Involved with Search NEU!\\" />
    <ModalContent className=\\"formModalContent\\">
      <Form as=\\"form\\">
        <div className=\\"feedbackParagraph\\">
          Thanks for your interest! We&#39;d love to have more people help out with the project. We are looking for people for both introductory level roles and leadership roles. There&#39;s a lot of CS stuff (new features, etc) and non-CS stuff (posters, marketing, outreach, etc) that we could work on, so it is no problem at all if you don&#39;t have a lot of experience in CS! Everything is flexible, and we could help you learn some programming along the way if you want to work on the siteÂ ðŸ™‚. Leave your name and some way we can get in contact (Facebook URL, email, Fortnite username, etc) and someone will reach out!
        </div>
        <TextArea name=\\"response\\" form=\\"feedbackForm\\" className=\\"feedbackTextbox\\" onChange={[Function: bound onTextAreaChange]} as=\\"textarea\\" rows={3} />
      </Form>
    </ModalContent>
    <ModalActions>
      <Button basic={true} color=\\"red\\" onClick={[Function: Function]} as=\\"button\\">
        <Icon name=\\"remove\\" as=\\"i\\" />
        Cancel
      </Button>
      <Button type=\\"submit\\" color=\\"green\\" form=\\"feedbackForm\\" onClick={[Function: bound onSubmit]} as=\\"button\\">
        <Icon name=\\"checkmark\\" as=\\"i\\" />
        Submit
      </Button>
    </ModalActions>
  </Modal>
</div>"
`;

exports[`should render the form 1`] = `
"<div className=\\"feedback-container\\">
  <Transition in={false} timeout={500} mountOnEnter={false} unmountOnExit={false} appear={false} enter={true} exit={true} onEnter={[Function: noop]} onEntering={[Function: noop]} onEntered={[Function: noop]} onExit={[Function: noop]} onExiting={[Function: noop]} onExited={[Function: noop]}>
    [function]
  </Transition>
  <Modal open={true} onClose={[Function: Function]} size=\\"small\\" className=\\"feedback-modal-container\\" centered={true} dimmer={true} closeOnDimmerClick={true} closeOnDocumentClick={false} eventPool=\\"Modal\\">
    <Header icon=\\"mail\\" content=\\"Get Involved with Search NEU!\\" />
    <ModalContent className=\\"formModalContent\\">
      <Form as=\\"form\\">
        <div className=\\"feedbackParagraph\\">
          Thanks for your interest! We&#39;d love to have more people help out with the project. We are looking for people for both introductory level roles and leadership roles. There&#39;s a lot of CS stuff (new features, etc) and non-CS stuff (posters, marketing, outreach, etc) that we could work on, so it is no problem at all if you don&#39;t have a lot of experience in CS! Everything is flexible, and we could help you learn some programming along the way if you want to work on the siteÂ ðŸ™‚. Leave your name and some way we can get in contact (Facebook URL, email, Fortnite username, etc) and someone will reach out!
        </div>
        <TextArea name=\\"response\\" form=\\"feedbackForm\\" className=\\"feedbackTextbox\\" onChange={[Function: bound onTextAreaChange]} as=\\"textarea\\" rows={3} />
      </Form>
    </ModalContent>
    <ModalActions>
      <Button basic={true} color=\\"red\\" onClick={[Function: Function]} as=\\"button\\">
        <Icon name=\\"remove\\" as=\\"i\\" />
        Cancel
      </Button>
      <Button type=\\"submit\\" color=\\"green\\" form=\\"feedbackForm\\" onClick={[Function: bound onSubmit]} as=\\"button\\">
        <Icon name=\\"checkmark\\" as=\\"i\\" />
        Submit
      </Button>
    </ModalActions>
  </Modal>
</div>"
`;
